scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryNetwork.scala

```diff
@@ -0,0 +1,376 @@
+package io.iohk.scalanet.discovery.ethereum.v4
+
+import cats.implicits._
+import cats.effect.{Clock}
+import cats.effect.concurrent.Deferred
+import com.typesafe.scalalogging.LazyLogging
+import io.iohk.scalanet.discovery.crypto.{PrivateKey, PublicKey, SigAlg}
+import io.iohk.scalanet.discovery.ethereum.Node
+import io.iohk.scalanet.discovery.hash.Hash
+import io.iohk.scalanet.peergroup.PeerGroup
+import io.iohk.scalanet.peergroup.PeerGroup.ServerEvent.ChannelCreated
+import io.iohk.scalanet.peergroup.Channel
+import io.iohk.scalanet.peergroup.Channel.{MessageReceived, DecodingError, UnexpectedError}
+import java.util.concurrent.TimeoutException
+import java.net.InetAddress
+import monix.eval.Task
+import monix.tail.Iterant
+import monix.catnap.CancelableF
+import scala.concurrent.duration._
+import scodec.{Codec, Attempt}
+import scala.util.control.{NonFatal, NoStackTrace}
+import scodec.bits.BitVector
+import io.iohk.scalanet.discovery.ethereum.v4.Payload.Neighbors
+
+/** Present a stateless facade implementing the RPC methods
+  * that correspond to the discovery protocol messages on top
+  * of the peer group representing the other nodes.
+  */
+trait DiscoveryNetwork[A] extends DiscoveryRPC[A] {
+
+  /** Start handling incoming requests using the local RPC interface.
+    * The remote side is identified by its ID and address.*/
+  def startHandling(handler: DiscoveryRPC[(PublicKey, A)]): Task[CancelableF[Task]]
+}
+
+object DiscoveryNetwork {
+
+  // Errors that stop the processing of incoming messages on a channel.
+  class PacketException(message: String) extends Exception(message) with NoStackTrace
+
+  def apply[A](
+      peerGroup: PeerGroup[A, Packet],
+      privateKey: PrivateKey,
+      toNodeAddress: A => Node.Address,
+      messageExpiration: FiniteDuration = 60.seconds,
+      // Timeout for individual requests.
+      requestTimeout: FiniteDuration = 3.seconds,
+      // Timeout for collecting multiple potential Neighbors responses.
+      kademliaTimeout: FiniteDuration = 7.seconds,
+      // Max number of neighbours to expect.
+      kademliaBucketSize: Int = 16
+  )(implicit codec: Codec[Payload], sigalg: SigAlg, clock: Clock[Task]): Task[DiscoveryNetwork[A]] = Task {
+    new DiscoveryNetwork[A] with LazyLogging {
+
+      import DiscoveryRPC.ENRSeq
+      import Payload._
+
+      private val expirationMillis = messageExpiration.toMillis
+
+      private val currentTimeMillis = clock.realTime(MILLISECONDS)
+
+      private val maxNeighborsPerPacket = getMaxNeighborsPerPacket
+
+      private val localNodeAddress = toNodeAddress(peerGroup.processAddress)
```

Comment:
I am not sure if this should be taken out of `peerGroup.processAddress` as in most cases this returns actually interface with which the peergroup has been started not acutual external ip address i.e most probably `127.0.0.1` in test envs or `0.0.0.0` in real working enviroment. And in ping message we probably want to use real ip address (although i am not sure that important here, as we do not respond to advertised address but to the address we got packet form)

Reply:
The concept of the `processAddress` is a bit nebulous to me. I saw that currently the `*PeerGroup.Config` classes basically use the `bindAddress` as the `processAddress`, but form the interface it looks like that its type `A` is actually addressable, otherwise how would `PeerGroup.client(to: A)` work? I thought actually `bindAddress` might be the one pointing at `0.0.0.0` and the `processAddress` would be the one you use to participate in the p2p network, the outwardly meaningful one. It seemed to make sense that if we use addresses of type `A` to connect to each other then these `A` addresses are what need to be sent to each other as well.

It will be important that this is correct in `Neighbor` packets, but you are right the ones in `Ping` are completely ignored. In the next PR I'm passing in a `Node` to the service, so we can do that here too if necessary.

Reply:
yep maybe we can think about it next. For now we have exteranal address discovery implemented on node level, maybe it would be worthwhile to move it scalanet ? then peer groups would be able to privde bind address and real external address ? just something to think about.

Reply:
Yes I think that component about opening firewall holes and discovering external addresses would be a great addition to Scalanet, it's not something I'd know how to do on my own, would be great if the library catered for it, great idea :+1:

Reply:
I will create a task to migrate those thing from our other project. There we got external address discovery, and upnp to automatic port forwarding in case user is behind the nat.

Reply:
I changed this in https://github.com/input-output-hk/scalanet/pull/96/commits/5e9e33f56cfad26381aceb1236c7c632568b9c57 to passthe `localNodeAddress` directly; `besu` seems to be [picking](https://github.com/hyperledger/besu/blob/master/ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/PeerDiscoveryAgent.java#L190) the TCP port from it.